data/src/main/java/com/microsoft/azure/kusto/data/KustoResultMapper.java

```diff
@@ -0,0 +1,106 @@
+package com.microsoft.azure.kusto.data;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+/**
+ * A class for mapping Kusto results to a list of pojos
+ * 
+ * @param <R>
+ *            pojo type returned by the mapping
+ */
+public class KustoResultMapper<R> {
+	
+	public static class Builder<R> {
+		final List<ObjectPopulator<R, ?, ?>>	queryResultColumns	= new ArrayList<>();
+		final Supplier<R>									objConstructor;
+		
+		public Builder(Supplier<R> objConstructor) {
+			this.objConstructor = objConstructor;
+		}
+		
+		<C> Builder<R> addColumn(KustoType<C> type, String name, boolean isNullable, BiConsumer<R, C> setter) {
+			this.queryResultColumns.add(ObjectPopulator.of(name, type, isNullable, setter));
+			return this;
+		}
+		
+		<C> Builder<R> addColumn(KustoType<C> type, String name, int ordinal, boolean isNullable, BiConsumer<R, C> setter) {
+			this.queryResultColumns.add(ObjectPopulator.of(name, ordinal, type, isNullable, setter));
+			return this;
+		}
+		
+		<C> Builder<R> addColumn(KustoType<C> type, int ordinal, boolean isNullable, BiConsumer<R, C> setter) {
+			this.queryResultColumns.add(ObjectPopulator.of(ordinal, type, isNullable, setter));
+			return this;
+		}
+		
+		public <C> Builder<R> addNonNullableColumn(KustoType<C> type, String name, BiConsumer<R, C> setter) {
+			return addColumn(type, name, false, setter);
+		}
+		
+		public <C> Builder<R> addNonNullableColumn(KustoType<C> type, int ordinal, BiConsumer<R, C> setter) {
+			return addColumn(type, ordinal, false, setter);
+		}
+		
+		public <C> Builder<R> addNonNullableColumn(KustoType<C> type, String name, int ordinal, BiConsumer<R, C> setter) {
+			return addColumn(type, name, ordinal, false, setter);
+		}
+		
+		public <C> Builder<R> addNullableColumn(KustoType<C> type, String name, BiConsumer<R, C> setter) {
+			return addColumn(type, name, true, setter);
+		}
+		
+		public <C> Builder<R> addNullableColumn(KustoType<C> type, String name, int ordinal, BiConsumer<R, C> setter) {
+			return addColumn(type, name, ordinal, true, setter);
+		}
+		
+		public <C> Builder<R> addNullableColumn(KustoType<C> type, int ordinal, BiConsumer<R, C> setter) {
```

Comment:
These 6 helper methods seem unnecessarily divided by whether the column is nullable. Why not accept "isNullable" as a parameter?

Reply:
> Why not accept "isNullable" as a parameter?

Just a personal preference. It makes the API more readable compared to a boolean flag that's easier to miss or just forget what it means. Consider the following two snippets:

1.

	static final KustoResultMapper<MyPojo> mapper = KustoResultMapper.newBuilder(MyPojo::new)
			.addNonNullableColumn(KustoType.INTEGER, "Id", MyPojo::setId)
			.addNullableColumn(KustoType.STRING, "Name", MyPojo::setName)
			.addNullableColumn(KustoType.DATETIME_ZONED_DATE_TIME, "TimeColumnName", MyPojo::setTime)
			.addNullableColumn(KustoType.GUID_UUID, "Guid", MyPojo::setUuid)
			.addNullableColumn(KustoType.STRING, "Enum", (r,t) -> r.setSomeEnum(SomeEnum.valueOf(t)))
			.build();

2.

	static final KustoResultMapper<MyPojo> mapper = KustoResultMapper.newBuilder(MyPojo::new)
			.addColumn(KustoType.INTEGER, "Id", MyPojo::setId, false)
			.addColumn(KustoType.STRING, "Name", MyPojo::setName, true)
			.addColumn(KustoType.DATETIME_ZONED_DATE_TIME, "TimeColumnName", MyPojo::setTime, true)
			.addColumn(KustoType.GUID_UUID, "Guid", MyPojo::setUuid, true)
			.addColumn(KustoType.STRING, "Enum", (r,t) -> r.setSomeEnum(SomeEnum.valueOf(t)), true)
			.build();

I tend to prefer (1), but have no strong feelings about it. I can change it to (2) if you prefer.

Reply:
You have a fair point, but I definitely prefer passing it as a mandatory (so it won't be missed) parameter - modern IDEs will show the fieldname. I think adding so many extra methods makes it more complicated.

Reply:
Done.